### ISSUE TITLE
EVAL does not handle operator computation properly

### ISSUE THREAD

**Issue by HeikoBecker** [HOL-Theorem-Prover/HOL#1050]
The functions in `computeLib` do not properly deal with overloaded operators.
It seems that they operator definition is inlined for definitions and theorems while it should be possible to change the evaluation scheme on the fly.
The following MWE illustrates the issue:

```
open realLib RealArith realTheory;

Definition list_mul_cst_def[nocompute]:
  list_mul_cst (c1:real) [] = [] ∧
  list_mul_cst c1 (c2::cs) = (c1 * c2) :: list_mul_cst c1 cs
End

Theorem list_mul_cst_comp[compute]:
  list_mul_cst c1 cs =
  if cs = [] then [] else
  let hd_comp = c1 * (HD cs) in hd_comp :: list_mul_cst c1 (TL cs)
Proof
  Cases_on 'cs' >> gs[list_mul_cst_def]
QED
(** now list_mul_cst_comp has bound * to the old(!) definition of real_mul/$*:real->real->real **)

fun dumb_conv t =
  let val _ = print "dumb_conv" in Q.prove('^t = 0', cheat) end

val _ = computeLib.del_consts["($*):real->real->real"]
val _ = computeLib.add_convs[("($*):real->real->real",2,dumb_conv)]

(** Does print "dumb_conv", i.e. it uses the conversion **)
val t1 = EVAL "2 * 3:real"

(** Does not print "dumb_conv" , i.e. resorts to the old evaluation scheme and does not use the new conversion **)
val t2 = EVAL "list_mul_cst 230 [290579; 29095789; 1029478]"

(** The correct theorem is in the compute set though **)
val tms = computeLib.listItems (computeLib.the_compset)
            |> List.filter(fn ((a,b),c) => a = "list_mul_cst")

(** If we add and remove it, the inlining seems to happen again **)
val _ = computeLib.del_funs[list_mul_cst_def]
val _ = computeLib.add_funs[list_mul_cst_comp]

(** Does print "dumb_conv", i.e. the conversion is triggered **)
val t3 = EVAL "list_mul_cst 230 [290579; 29095789; 1029478]"
```

**Comment by mn200** [comment]
The making of the definition `nocompute` and then using a separate theorem has nothing to do with the issue.  Nor does it have anything to do with overloading. 

It is true that the rewrite to use for embedded terms within a rewrite is effectively inlined, and this is done with references in a horrible way that bypasses database updates .  This suggests that if you want to override `EVAL`’s treatment of multiplication you should do that before you add constants that want to use multiplication. 

In other words, update to use fancy multiplication, and then make the definition. In order to get this to happen right in exported theory files, you will need to call `Theory.add_ML_dependency` to ensure that the module with the conversion definition and installation gets loaded before theory defining `list_mul_cst` does.

**Comment by HeikoBecker** [comment]
Thanks for the explanation. I have now a working version that permanently installs the conversion for all dependent theories. I am leaving the fixed MWE here for reference:

```
structure mweLib = struct 

open realLib RealArith realTheory;

Definition list_mul_cst_def:
  list_mul_cst (c1:real) [] = [] ∧
  list_mul_cst c1 (c2::cs) = (c1 * c2) :: list_mul_cst c1 cs
End

fun dumb_conv t =
  let val _ = print "dumb_conv" in Q.prove('^t = 0', cheat) end

fun use_dumb_conv() =
    (Theory.add_ML_dependency "mweLib";
      Theory.adjoin_to_theory
        {sig_ps = NONE,
          struct_ps =
            SOME(fn _ => PP.add_string "val _ = mweLib.use_dumb_conv();")};
      (* Remove old definitions of multiplication for reals and nums *)
      computeLib.del_consts [“($*):real->real->real”];
      (* Add karatsuba multiplication as conversion for reals and nums *)
      computeLib.add_convs [(“($*):real->real->real”,2, dumb_conv)])
end
```
To use `dumb_conv` for evaluation, it would suffice to call `mweLib.use_dumb_conv()` in the theory that defines the HOL4 functions that should use the conversion when evaluated with `EVAL`.
